File: Python/merge_sort.py
def merge_two_sorted_lists(right,left):
    r_idx = l_idx = 0
    sorted_lst = []

    while r_idx <= len(right)-1 and l_idx <= len(left)-1:
        if right[r_idx] <= left[l_idx]:
            sorted_lst.append(right[r_idx])
            r_idx = r_idx + 1
        else:
            sorted_lst.append(left[l_idx])
            l_idx = l_idx + 1

    while r_idx <= len(right)-1:
        sorted_lst.append(right[r_idx])
        r_idx = r_idx + 1

    while l_idx <= len(left)-1:
        sorted_lst.append(left[l_idx])
        l_idx = l_idx + 1
    
    return sorted_lst

def merge_sort(lst):
    if len(lst) <= 1:
        return lst
    else:
        mid = len(lst) // 2
        right = merge_sort(lst[mid:])
        left = merge_sort(lst[:mid])
        return merge_two_sorted_lists(right,left)

lst1 = [1,6,4,10,3,8]
# lst2 = [2,5,8]

print(merge_sort(lst1))

File: C++/merge_sort.cpp
#include <iostream>
using namespace std;
#include <vector>


vector<int> merge_two_sorted_lists(vector<int> left , vector<int> right){
    int r_ind = 0;
    int l_ind = 0;
    vector<int> sorted_list = {};
    
    while ((r_ind < right.size()) && (l_ind < left.size())){
        if (right[r_ind] <= left[l_ind]){
            sorted_list.push_back(right[r_ind]);
            r_ind += 1;
        }
        else{
            sorted_list.push_back(left[l_ind]);
            l_ind += 1;
        }
    }

    while (r_ind < right.size()){
        sorted_list.push_back(right[r_ind]);
        r_ind += 1;
    }

    while (l_ind < left.size()){
        sorted_list.push_back(left[l_ind]);
        l_ind += 1;
    }

    return sorted_list;
};

vector<int> merge_sort(vector<int> nums){
    // if nums is empty or has one element, return nums
    if (nums.size() <= 1){
        return nums;
    }

    // divide nums to left and right
    vector<int> left;
    vector<int> right;

    int lgth = nums.size();
    left.insert(left.end() , nums.begin() , nums.begin() + (lgth/2));
    right.insert(right.end() , nums.begin() + (lgth/2) , nums.end());
    
    // do merge sort on left and right
    left = merge_sort(left);
    right = merge_sort(right);
    
    // return merge_two_sorted_lists(left,right)
    vector<int> ans = merge_two_sorted_lists(left,right);
    return ans;
};


int main(){
    
    vector<int> nums = {1,5,4,10,7,3};          // Original nums
    vector<int> nums_sorted = {1,3,4,5,7,10};   // Sorted nums
    
    bool test = merge_sort(nums) == nums_sorted;
    if (test){
        cout << "Sorted Successfully" << endl;
    }
    else{
        cout << "Failed" << endl;
    }

    return 0;
}